finality-aleph/src/validator_network/manager/mod.rs

```diff
@@ -30,191 +41,165 @@ impl Display for SendError {
+                    write!(
+                        f,
+                        "missing - {:?} [{}]; ",
+                        self.missing_incoming.len(),
+                        pretty_authority_id_set(&self.missing_incoming),
+                    )?;
+                }
+            }
         }
 
-        if !self.missing_peers.is_empty() {
-            let peers = self
-                .missing_peers
-                .iter()
-                .map(|authority_id| authority_id.to_short_string())
-                .collect::<Vec<_>>()
-                .join(", ");
-            write!(f, "missing - {:?} [{}];", self.missing_peers.len(), peers)?;
+        match wanted_outgoing {
+            0 => write!(f, "not attempting any outgoing connections; ")?,
+            _ => {
+                write!(f, "attempting {:?} outgoing connections; ", wanted_outgoing)?;
+                if !self.outgoing_peers.is_empty() {
+                    write!(
+                        f,
+                        "have - {:?} [{}]; ",
+                        self.incoming_peers.len(),
+                        pretty_authority_id_set(&self.outgoing_peers),
+                    )?;
+                }
+                if !self.missing_outgoing.is_empty() {
+                    write!(
+                        f,
+                        "missing - {:?} [{}]; ",
+                        self.missing_incoming.len(),
+                        pretty_authority_id_set(&self.missing_outgoing),
+                    )?;
+                }
+            }
         }
 
         Ok(())
     }
 }
 
+/// Possible results of adding connections.
+#[derive(Debug, PartialEq, Eq)]
+pub enum AddResult {
+    /// We do not want to maintain a connection with this peer.
+    Uninterested,
+    /// Connection added.
+    Added,
+    /// Old connection replaced with new one.
+    Replaced,
+}
+
 impl<A: Data, D: Data> Manager<A, D> {
     /// Create a new Manager with empty list of peers.
-    pub fn new() -> Self {
+    pub fn new(own_id: AuthorityId) -> Self {
         Manager {
-            addresses: HashMap::new(),
-            outgoing: HashMap::new(),
-            incoming: HashMap::new(),
+            wanted: DirectedPeers::new(own_id),
+            have: HashMap::new(),
         }
     }
 
+    fn active_connection(&self, peer_id: &AuthorityId) -> bool {
+        self.have
+            .get(peer_id)
+            .map(|sender| !sender.is_closed())
+            .unwrap_or(false)
+    }
+
     /// Add a peer to the list of peers we want to stay connected to, or
     /// update the list of addresses if the peer was already added.
-    /// Returns whether this peer is a new peer.
+    /// Returns whether we should start attempts at connecting with the peer.
     pub fn add_peer(&mut self, peer_id: AuthorityId, addresses: Vec<A>) -> bool {
-        self.addresses.insert(peer_id, addresses).is_none()
+        self.wanted.add_peer(peer_id, addresses)
```

Comment:
What if old `addresses` is equal to the old set and we already hold connection in `self.have`? We will reconnect unnecessarily, right?

Reply:
oh, I see in the testcase that it is checked - within `self.wanted.add_peer()`, true? then maybe please add a word about it in docs

Reply:
Expanded the last sentence in the docs of this method, but I'm not sure that's what you meant, please let me know if more clarification is required.